Roslyn fails to enforce definitely assigned rules and cannot emit code for local functions with closures over pattern variables in switch statements

Following code should have error CS0165: Use of unassigned local variable 'x'

```cs
class Program
{
    static void Main(object[] args)
    {
        switch(args[0])
        {
            case string x:
                Foo(); // missing error
                break;
            case int x:
                void Foo() => System.Console.WriteLine(x);
                break;
        }
    }
}
```

Roslyn doesn't detect the error and fails to compile the code, this is VS2017 RC output:
```
csc.exe exited with code 255
```

Note that if you reorder the cases you get correct error i.e.
```cs
class Program
{
    static void Main(object[] args)
    {
        switch(args[0])
        {
            case int x:
                void Foo() => System.Console.WriteLine(x);
                break;
            case string x:
                Foo(); // error CS0165: Use of unassigned local variable 'x'
                break;
        }
    }
}
```

TryRoslyn emits following stacktrace for original code (it may be a bit outdated though):
```
The given key was not present in the dictionary. 
at System.Collections.Generic.Dictionary`2.get_Item(TKey key) 
at Microsoft.CodeAnalysis.CodeGen.LocalSlotManager.GetLocal(ILocalSymbol symbol) in ...\Src\Compilers\Core\Portable\CodeGen\LocalSlotManager.cs:line 148 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.GetLocal(LocalSymbol symbol) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitStatement.cs:line 1387 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.GetLocal(BoundLocal localExpression) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitStatement.cs:line 1382 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitLocalAddress(BoundLocal localAccess) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitAddress.cs:line 167 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitAddress(BoundExpression expression, AddressKind addressKind) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitAddress.cs:line 38 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitArgument(BoundExpression argument, RefKind refKind) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitExpression.cs:line 594 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitArguments(ImmutableArray`1 arguments, ImmutableArray`1 parameters) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitExpression.cs:line 769 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitCallExpression(BoundCall call, UseKind useKind) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitExpression.cs:line 1480 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitExpressionCore(BoundExpression expression, Boolean used) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitExpression.cs:line 92 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitExpressionCoreWithStackGuard(BoundExpression expression, Boolean used) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitExpression.cs:line 72 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitExpression(BoundExpression expression, Boolean used) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitExpression.cs:line 60 
```

ping @agocke

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs
@@ -92,9 +92,9 @@ namespace Microsoft.CodeAnalysis.CSharp
                     if (slot > 0 && !this.State.IsAssigned(slot))
                     {
                         // Local functions can "call forward" to after a variable has
-                        // been defined but before it has been assigned, so we can never
-                        // consider the definition location when reporting errors.
-                        ReportUnassigned(symbol, node, slot, considerDefineLocation: false);
+                        // been declared but before it has been assigned, so we can never
+                        // consider the declaration location when reporting errors.
+                        ReportUnassigned(symbol, node, slot, skipIfUseBeforeDeclaration: false);
                     }
                 }
             }

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs
@@ -61,7 +61,41 @@ namespace Microsoft.CodeAnalysis.CSharp
                 if (reads[slot])
                 {
                     var symbol = variableBySlot[slot].Symbol;
-                    CheckAssigned(symbol, syntax, slot);
+                    CheckIfAssignedDuringLocalFunctionReplay(symbol, syntax, slot);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Check that the given variable is definitely assigned when replaying local function
+        /// reads. If not, produce an error.
+        /// </summary>
+        /// <remarks>
+        /// Specifying the slot manually may be necessary if the symbol is a field,
+        /// in which case <see cref="VariableSlot(Symbol, int)"/> will not know
+        /// which containing slot to look for.
+        /// </remarks>
+        private void CheckIfAssignedDuringLocalFunctionReplay(Symbol symbol, SyntaxNode node, int slot)
+        {
+            Debug.Assert(!IsConditionalState);
+            if ((object)symbol != null)
+            {
+                NoteRead(symbol);
+
+                if (this.State.Reachable)
+                {
+                    if (slot >= this.State.Assigned.Capacity)
+                    {
+                        Normalize(ref this.State);
+                    }
+
+                    if (slot > 0 && !this.State.IsAssigned(slot))
+                    {
+                        // Local functions can "call forward" to after a variable has
+                        // been defined but before it has been assigned, so we can never
+                        // consider the definition location when reporting errors.
+                        ReportUnassigned(symbol, node, slot, considerDefineLocation: false);
+                    }
                 }
             }
         }

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs
@@ -896,37 +896,15 @@ namespace Microsoft.CodeAnalysis.CSharp
             }
         }
 
-        /// <summary>
-        /// Check that the given variable is definitely assigned.  If not, produce an error.
-        /// </summary>
-        /// <remarks>
-        /// Specifying the slot manually may be necessary if the symbol is a field,
-        /// in which case <see cref="VariableSlot(Symbol, int)"/> will not know
-        /// which containing slot to look for.
-        /// </remarks>
-        private void CheckAssigned(Symbol symbol, SyntaxNode node, int slot)
-        {
-            Debug.Assert(!IsConditionalState);
-            if ((object)symbol != null)
-            {
-                NoteRead(symbol);
-
-                if (this.State.Reachable)
-                {
-                    if (slot >= this.State.Assigned.Capacity)
-                    {
-                        Normalize(ref this.State);
-                    }
-
-                    if (slot > 0 && !this.State.IsAssigned(slot))
-                    {
-                        ReportUnassigned(symbol, node, slot);
-                    }
-                }
-            }
-        }
-
-        private void ReportUnassigned(Symbol symbol, SyntaxNode node, int? slotOpt)
+        /// <param name="symbol">Symbol to variable that is unassigned.</param>
+        /// <param name="node">Syntax where read occurs.</param>
+        /// <param name="slotOpt">Optional slot where variable is located.</param>
+        /// <param name="considerDefineLocation">
+        /// True if error reporting should consider the location where the
+        /// variable is defined (for instance, eliding errors about reading
+        /// variables that have not yet been defined).
+        /// </param>
+        private void ReportUnassigned(Symbol symbol, SyntaxNode node, int? slotOpt, bool considerDefineLocation = true)
         {
             int slot = slotOpt ?? VariableSlot(symbol);
             if (slot <= 0) return;
@@ -949,7 +927,14 @@ namespace Microsoft.CodeAnalysis.CSharp
             }
 
             if (slot >= _alreadyReported.Capacity) _alreadyReported.EnsureCapacity(nextVariableSlot);
-            if (!_alreadyReported[slot] && VariableType(symbol)?.IsErrorType() != true)
+            if (considerDefineLocation &&
+                symbol.Kind == SymbolKind.Local &&
+                (symbol.Locations.Length == 0 || node.Span.End < symbol.Locations[0].SourceSpan.Start))
+            {
+                // We've already reported the use of a local before its declaration.  No need to emit
+                // another diagnostic for the same issue.
+            }
+            else if (!_alreadyReported[slot] && VariableType(symbol)?.IsErrorType() != true)
             {
                 // CONSIDER: could suppress this diagnostic in cases where the local was declared in a using
                 // or fixed statement because there's a special error code for not initializing those.

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs
@@ -899,12 +899,12 @@ namespace Microsoft.CodeAnalysis.CSharp
         /// <param name="symbol">Symbol to variable that is unassigned.</param>
         /// <param name="node">Syntax where read occurs.</param>
         /// <param name="slotOpt">Optional slot where variable is located.</param>
-        /// <param name="considerDefineLocation">
+        /// <param name="skipIfUseBeforeDeclaration">
         /// True if error reporting should consider the location where the
-        /// variable is defined (for instance, eliding errors about reading
-        /// variables that have not yet been defined).
+        /// variable is declared (for instance, eliding errors about reading
+        /// variables that have not yet been declared).
         /// </param>
-        private void ReportUnassigned(Symbol symbol, SyntaxNode node, int? slotOpt, bool considerDefineLocation = true)
+        private void ReportUnassigned(Symbol symbol, SyntaxNode node, int? slotOpt, bool skipIfUseBeforeDeclaration = true)
         {
             int slot = slotOpt ?? VariableSlot(symbol);
             if (slot <= 0) return;
@@ -927,7 +927,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             }
 
             if (slot >= _alreadyReported.Capacity) _alreadyReported.EnsureCapacity(nextVariableSlot);
-            if (considerDefineLocation &&
+            if (skipIfUseBeforeDeclaration &&
                 symbol.Kind == SymbolKind.Local &&
                 (symbol.Locations.Length == 0 || node.Span.End < symbol.Locations[0].SourceSpan.Start))
             {

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs
@@ -949,12 +949,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             }
 
             if (slot >= _alreadyReported.Capacity) _alreadyReported.EnsureCapacity(nextVariableSlot);
-            if (symbol.Kind == SymbolKind.Local && (symbol.Locations.Length == 0 || node.Span.End < symbol.Locations[0].SourceSpan.Start))
-            {
-                // We've already reported the use of a local before its declaration.  No need to emit
-                // another diagnostic for the same issue.
-            }
-            else if (!_alreadyReported[slot] && VariableType(symbol)?.IsErrorType() != true)
+            if (!_alreadyReported[slot] && VariableType(symbol)?.IsErrorType() != true)
             {
                 // CONSIDER: could suppress this diagnostic in cases where the local was declared in a using
                 // or fixed statement because there's a special error code for not initializing those.

--- a/src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs
+++ b/src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs
@@ -3072,6 +3072,20 @@ class Program
             VerifyOutput(source, "2", TestOptions.ReleaseExe.WithAllowUnsafe(true));
         }
 
+        [Fact]
+        [WorkItem(15322, "https://github.com/dotnet/roslyn/issues/15322")]
+        public void UseBeforeDeclaration()
+        {
+            var src = @"
+Assign();
+Local();
+int x;
+void Local() => System.Console.WriteLine(x);
+void Assign() { x = 5; }";
+
+            VerifyOutputInMain(src, "5");
+        }
+
         internal CompilationVerifier VerifyOutput(string source, string output, CSharpCompilationOptions options)
         {
             var comp = CreateCompilationWithMscorlib45AndCSruntime(source, options: options);

--- a/src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs
+++ b/src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs
@@ -1091,5 +1091,51 @@ class C
 }");
             comp.VerifyDiagnostics();
         }
+
+        [Fact]
+        [WorkItem(15298, "https://github.com/dotnet/roslyn/issues/15298")]
+        public void UnassignedUndefinedVariable()
+        {
+            var comp = CreateCompilationWithMscorlib(@"
+class C
+{
+    void M()
+    {
+        Foo();
+        int x;
+        void Foo() => x++;
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (6,9): error CS0165: Use of unassigned local variable 'x'
+                //         Foo();
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "Foo()").WithArguments("x").WithLocation(6, 9));
+        }
+
+        [Fact]
+        [WorkItem(15322, "https://github.com/dotnet/roslyn/issues/15322")]
+        public void UseBeforeDeclarationInSwitch()
+        {
+            var comp = CreateCompilationWithMscorlib(@"
+class Program
+{
+    static void Main(object[] args)
+    {
+        switch(args[0])
+        {
+            case string x:
+                Foo(); 
+                break;
+            case int x:
+                void Foo() => System.Console.WriteLine(x);
+                break;
+        }
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (9,17): error CS0165: Use of unassigned local variable 'x'
+                //                 Foo();
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "Foo()").WithArguments("x").WithLocation(9, 17));
+        }
     }
 }

--- a/src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs
+++ b/src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs
@@ -1101,15 +1101,25 @@ class C
 {
     void M()
     {
-        Foo();
-        int x;
-        void Foo() => x++;
+        {
+            Foo();
+            int x = 0;
+            void Foo() => x++;
+        }
+        {
+            System.Action a = Foo;
+            int x = 0;
+            void Foo() => x++;
+        }
     }
 }");
             comp.VerifyDiagnostics(
-                // (6,9): error CS0165: Use of unassigned local variable 'x'
-                //         Foo();
-                Diagnostic(ErrorCode.ERR_UseDefViolation, "Foo()").WithArguments("x").WithLocation(6, 9));
+                // (7,13): error CS0165: Use of unassigned local variable 'x'
+                //             Foo();
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "Foo()").WithArguments("x").WithLocation(7, 13),
+                // (12,31): error CS0165: Use of unassigned local variable 'x'
+                //             System.Action a = Foo;
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "Foo").WithArguments("x").WithLocation(12, 31));
         }
 
         [Fact]

